eventsourcing/src/main/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplier.java

```diff
@@ -0,0 +1,117 @@
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull Class<M> modelType) {
+        this(
+                modelType,
+                AnnotatedHandlerInspector.inspectType(modelType,
+                                                      ClasspathParameterResolverFactory.forClass(modelType),
+                                                      ClasspathHandlerDefinition.forClass(modelType))
+        );
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param modelType The type of model this instance will handle state changes for.
+     * @param inspector The inspector to use to find the annotated handlers on the model.
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull Class<M> modelType,
+                                            @Nonnull AnnotatedHandlerInspector<M> inspector
+    ) {
+        requireNonNull(modelType, "Model Type may not be null");
+        this.inspector = requireNonNull(inspector,
+                                        "Annotated Handler Inspector may not be null");
+    }
+
+    @Override
+    public M apply(@Nonnull M model, @Nonnull EventMessage<?> event, @Nonnull ProcessingContext processingContext) {
+        requireNonNull(model, "Model may not be null");
+        requireNonNull(event, "Event Message may not be null");
+
+        try {
+            var eventHandlerResult = handle(model, event, processingContext).join();
+            return modelFromStreamResultOrUpdatedExisting(eventHandlerResult, model);
+        } catch (Exception e) {
+            throw new StateEvolvingException(
+                    "Failed to apply event [" + event.type() + "] in order to evolve [" + model.getClass() + "] state",
+                    e);
+        }
+    }
+
+    private M modelFromStreamResultOrUpdatedExisting(MessageStream.Entry<?> potentialModelFromStream, M existing) {
+        if (potentialModelFromStream != null) {
+            var resultPayload = potentialModelFromStream.message().getPayload();
+            if (resultPayload != null && existing.getClass().isAssignableFrom(resultPayload.getClass())) {
+                //noinspection unchecked
+                return (M) existing.getClass().cast(resultPayload);
+            }
+        }
+        return existing;
+    }
+
+    private CompletableFuture<? extends MessageStream.Entry<?>> handle(
+            M model,
+            EventMessage<?> event,
+            ProcessingContext processingContext
+    ) {
+        var listenerType = model.getClass();
+        var handler =
+                inspector.getHandlers(listenerType)
+                         .filter(h -> h.canHandle(event, processingContext))
+                         .findFirst();
+        if (handler.isPresent()) {
+            var interceptor = inspector.chainedInterceptor(listenerType);
+            var stream = interceptor.handle(event, processingContext, model, handler.get());
+            return stream.firstAsCompletableFuture();
```

Comment:
This will no long work once you merge the af-5.0 with the new Stream interfaces into this branch.

Reply:
Will be `first().asCompletableFuture()`

Reply:
Ahhh of course! Yes, the evolve model would need a Stream.Single :thinking: